org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ProjectsManager.java

```diff
@@ -403,6 +408,111 @@ public void fileChanged(String uriString, CHANGE_TYPE changeType) {
 		}
 	}
 
+	private void refreshRootFolders(String uriString) {
+		if (!shouldImportProjects(uriString)) {
+			return;
+		}
+		FeatureStatus status = preferenceManager.getPreferences().getImportNewProjectsConfigurationStatus();
+		switch (status) {
+			case automatic:
+				Job job = new Job("Import Projects") {
+
+					@Override
+					protected IStatus run(IProgressMonitor monitor) {
+						importProjects(uriString);
+						return Status.OK_STATUS;
+					}
+
+					/* (non-Javadoc)
+					 * @see org.eclipse.core.runtime.jobs.Job#belongsTo(java.lang.Object)
+					 */
+					@SuppressWarnings("unchecked")
+					@Override
+					public boolean belongsTo(Object family) {
+						Collection<IPath> roots = preferenceManager.getPreferences().getRootPaths();
+						Collection<IPath> rootPathsSet = roots.stream().collect(Collectors.toSet());
+						boolean equalToRootPaths = false;
+						if (family instanceof Collection<?>) {
+							equalToRootPaths = rootPathsSet.equals(((Collection<IPath>) family).stream().collect(Collectors.toSet()));
+						}
+						return InitHandler.JAVA_LS_INITIALIZATION_JOBS.equals(family) || equalToRootPaths;
+					}
+
+				};
+				job.setSystem(true);
+				job.schedule(1000L);
+			case disabled:
+				break;
+			default:
+				if (client != null) {
+					String cmd = "java.importProjects.status";
+					TextDocumentIdentifier uri = new TextDocumentIdentifier(uriString);
+					ActionableNotification importNewProjectsConfigurationNotification = new ActionableNotification().withSeverity(MessageType.Info).withMessage("A new file was added. Do you want to import projects?").withCommands(
+							asList(new Command("Never", cmd, asList(uri, FeatureStatus.disabled)), new Command("Now", cmd, asList(uri, FeatureStatus.interactive)), new Command("Always", cmd, asList(uri, FeatureStatus.automatic))));
+					client.sendActionableNotification(importNewProjectsConfigurationNotification);
+				}
+		}
+	}
+
+	/**
+	 * <p>
+	 * <strong>NOTE:</strong> Use only for testing.
+	 * </p>
+	 */
+	public boolean shouldImportProjects(String uriString) {
+		IPath path = ResourceUtils.canonicalFilePathFromURI(uriString);
+		File file = path.toFile();
+		if (file.isFile()) {
+			File parent = file.getParentFile();
+			Collection<IPath> rootPaths = preferenceManager.getPreferences().getRootPaths();
+			for (IPath rootPath : rootPaths) {
+				if ((parent.getParentFile() != null && parent.getParentFile().equals(rootPath.toFile())) || parent.equals(rootPath.toFile())) {
```

Comment:
Hey @snjeza, 

Does that mean that if the project has a deep nest child module, we cannot handle that case? for example, a project which has following structure:
```
root
  --pom.xml
  --parent
    --pom.xml
  --module-a
    --pom.xml
    --parent-module-a
        --pom.xml
    --module-a-1
        --pom.xml
    --module-a-2
        --pom.xml
  --module-b
    --pom.xml
```

Reply:
See https://github.com/eclipse/eclipse.jdt.ls/pull/901#issuecomment-447515630